src/boot/store.js

```diff
@@ -217,6 +219,14 @@ const migrations: { [string]: (GlobalState) => GlobalState } = {
     narrows: Immutable.Map(state.narrows),
   }),
 
+  // Change format of keys representing narrows.
+  '17': state => ({
+    ...dropCache(state),
+    drafts: objectFromEntries(
+      Object.keys(state.drafts).map(key => [keyFromNarrow(JSON.parse(key)), state.drafts[key]]),
```

Comment:
In an ideal world, I guess, a migration would be set in stone as soon as it's committed. (That's partly why, e.g., 6797cdc1b spelled out its own function instead of calling `fixRealmUrl`.) If `keyFromNarrow` changes, and this migration still uses `keyFromNarrow`, some subtle bugs could be introduced. The number of affected users might be quite small, though—and it would be pretty verbose to clone `keyFromNarrow`'s current implementation for this migration, hmm.

Still, here's one plausibly realistic scenario.

I think we're hoping to change `stream:s:${name}` to something like `stream:n:${streamId}`, for #4333, right? We would do so by changing `keyFromNarrow`. That change would require its own migration, to get from the `:s:` form to the `:n:` form. But if that new migration rigidly expected (as would be very reasonable) to get the `:s:` form as input, some users would get crashes. In particular, those users who hadn't yet taken this first migration—which, because it uses (the changed) `keyFromNarrow`, would have taken them straight to the `:n:` form.

I'm not sure how many users might actually be affected (probably not a lot), but it might be kind of a pain to debug for those that are. What do you think?

Reply:
> Because duplicating the logic of `keyFromNarrow` here would be a pain, I think the happiest plan is to go ahead and invoke it. That will require extra care around migrations for any future changes to the key format; but if that's ever too much, an escape hatch is to just drop the drafts. We'll be doing a series of those future changes soon, so this will be fresh in mind (no risk of forgetting) and we can always revisit then; and this is an inherently pretty stable area, so once those are done, those will quite possibly be the last changes we ever make to it, apart from adding new narrow types.

OK, that seems sensible. 🙂